### ISSUE TITLE
When keyring is disabled, poetry tries to use it anyways. 

### ISSUE THREAD

**Issue by lakinwecker** [python-poetry/poetry#5250]
- [x] I am on the [latest](https://github.com/python-poetry/poetry/releases/latest) Poetry version.
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] If an exception occurs when executing a command, I executed it again in debug mode (`-vvv` option).

- **Arch linux up to date**: 
- **1.1.13**: 
- **Link of a [Gist](https://gist.github.com/) with the contents of your pyproject.toml file**: [pyproject.toml](https://gist.github.com/lakinwecker/f45a753fea8c5650fe6367c254e6af15)

## Issue
If you disable keyring via it's recommended route: https://github.com/jaraco/keyring#disabling-keyring you end up with a "null" keyring being used. This is a useless keyring, but poetry continues to try and use it regardless.   poetry should treat it the same as if keyring was unavailable.

**Comment by lakinwecker** [comment]
To be clear, when it fails this way, it fails to do things like it doesn't try and use the environment variable. It just silently ignores it.

**Comment by tyler-8** [comment]
> To be clear, when it fails this way, it fails to do things like it doesn't try and use the environment variable. It just silently ignores it.

This is a key symptom. I've been very confused as to why my correctly set environment variables for username/password of a private repo are not working. It wasn't until I disabled the keyring (as apparently I'd configured poetry to use it in the past) that I realized that truly the envars were being bypassed. Instead I get prompted for the password when doing `poetry publish -r myrepo`.

```
# envars
POETRY_HTTP_BASIC_MYREPO_USERNAME=someuser
POETRY_HTTP_BASIC_MYREPO_PASSWORD=password

# poetry config --list

repositories.myrepo.url = "https://internalpypi.domain.com/pypi"
```

**Comment by kadler** [comment]
I ran in to this same issue today. Looks like a similar check like this needs to be made for "null":

https://github.com/python-poetry/poetry/blob/master/src/poetry/utils/password_manager.py#L100-L102

**Comment by kadler** [comment]
As a workaround, the "fail" backend can be used: `export PYTHON_KEYRING_BACKEND=keyring.backends.fail.Keyring`

**Comment by lakinwecker** [comment]
> I ran in to this same issue today. Looks like a similar check like this needs to be made for "null":
> 
> https://github.com/python-poetry/poetry/blob/master/src/poetry/utils/password_manager.py#L100-L102

Yes, that's what the associated pull request does. :+1: Just waiting on me to have the time to get the unit tests done and I don't know when that will happen.

**Comment by kadler** [comment]
Oh sorry, I didn't see there was a PR already. Nice!

**Comment by epogrebnyak** [comment]
> As a workaround, the "fail" backend can be used: `export PYTHON_KEYRING_BACKEND=keyring.backends.fail.Keyring`

This is genius, helped so much, thank you.

**Comment by cal97g** [comment]
2023 and this issue still exists and is not fixed. :)

**Comment by lakinwecker** [comment]
This issue was fixed with the #5251 and released sometime thereafter. I haven't run into the issue since that point.  If you're running into something similar can you double check that you're using a recent enough version so that it includes the fix and if so, please submit details of how to reproduce it, rather than just claiming it still exists?

**Comment by maxbachmann** [comment]
I am running into this with 1.4.2 again as well. I just tried to run `poetry lock` in the cleo repository.

**Comment by lakinwecker** [comment]
I can't reproduce it: https://gist.github.com/lakinwecker/e85279c2fece8bec211c3608567c4162

Do you have an environment variable for `PYTHON_KEYRING_BACKEND` set? I don't
```-> env | grep KEYRING

cleo on  main is 📦 v2.1.0.dev0 via 🐍 v3.11.3 (cleo-py3.11) 
!-> ```

**Comment by aldum** [comment]
Question: why is it trying to access a keyring when installing dependencies?

**Comment by ralbertazzi** [comment]
https://github.com/python-poetry/poetry/issues/1917

**Comment by rdbisme** [comment]
> Question: why is it trying to access a keyring when installing dependencies?

Because sometimes you might want to access dependencies on an authenticated custom pypi registry?

**Comment by codethief** [comment]
Just ran into this again and took me a while to figure this out (again), since `poetry add -vvv <package>` just froze and didn't output anything useful. *sigh*

Can we please at least re-open this ticket?

**Comment by radoering** [comment]
There is #8623 and #8761. Do you think we need a third issue?

And there is #8910 (which will be included in 1.8). Apart from introducing a config setting to disable keyring, it refactored keyring access and thereby fixed some bugs.

**Comment by maciej-gol** [comment]
@radoering From my POV, this doesn't address my concerns, as I still believe keyring should be disabled by default and opt-in for users that _actually need_ it. That solution no different than putting `export PYTHON_KEYRING_BACKEND=keyring.backends.fail.Keyring` into my `.bashrc`. But its a step into good direction, nonetheless.

**Comment by ManBearPigg** [comment]
> There is #8623 and #8761. Do you think we need a third issue?

Prob because it breaks the whole project and the obscure fix is hard to find. I know it's free software I'm just saying why

**Comment by cal97g** [comment]
Let me help ya'll out with some pseudocode

```
if need_keyring:
  ask_for_keyring()
```

**Comment by maciej-gol** [comment]
@cal97g nice try, 2/10. Missing tests.

**Comment by alanwilter** [comment]
Why this issue is closed? Spent two days and two developers to figure this out.

`poetry` could be much smarter here. Could at least warn about it. This is another reason why we considering `hatch`.
